dimod/include/dimod/quadratic_model.h

```diff
@@ -252,9 +254,38 @@ class Neighborhood {
         }
     }
 
+    /// Request that the neighborhood capacity be at least enough to contain `n`
+    /// elements.
+    void reserve(index_type n) {
+        neighbors.reserve(n);
+        quadratic_biases.reserve(n);
+    }
+
     /// Return the size of the neighborhood.
     size_type size() const { return neighbors.size(); }
 
+    /// Sort the neighborhood and sum the biases of duplicate variables.
+    void sort_and_sum() {
+        if (!std::is_sorted(neighbors.begin(), neighbors.end())) {
+            utils::zip_sort(neighbors, quadratic_biases);
+        }
+
+        // now remove any duplicates, adding the biases
+        size_type i = 0;
+        while (i + 1 < size()) {
+            if (neighbors[i] == neighbors[i + 1]) {
```

Comment:
The better thing to do is to go allocate memory using malloc of the same size as the quadratic biases and and neighbors, write into the buffer the summarized results,  swap the current neighbors and quadratic_biases with empty vectors to ensure memory release( std::vector<index_type>().swap(neighbors) ) . Then use std::insert to insert from the allocated memory. 

Saves you from doing potentially n^2 work when there is a duplicate for each element.

Reply:
You could do other optimizations like go over the array once and see if there is any duplicate at all and only then go for the copying and allocating ...

Reply:
On second thoughts you do not need an extra array, you just need two indices and one temp neighbor storing the last seen neighbor.  one index pointing to the point you are at, one pointing upto the point you just stored the unique values    if duplicate add to the last pointed unique value, else increment the unique value counter and store to that point. In the end resize upto the uniquely pointed value. No extra memory overhead :)